### PR TITLE
quantum-espresso: ensure no space in HDF5 lib variable

### PR DESCRIPTION
**Pull request by etiennemlb**
Since 6.6 some awk black magic is used to parse the output of the HDF5 wrappers to get hdf5 libs (https://github.com/QEF/q-e/commit/5fb1195b0844e1052b7601f18ab5c700f9cbe648#diff-66930cefe8d40faeb19acef21b7e1c13a81822c6e3adfb70496aef7ec1fb1865).

This awk magic fails at least when you use the hdf5 wrappers on my produced by spack, using gcc 13 on my machine. The issue is like so:
```
$ <spack_dir>p -show
.../gcc-13.2.0/hdf5-1.14.3-u6yn/h5pfc
dir is .../gcc-13.2.0/hdf5-1.14.3-u6yn
gfortran .../gcc-13.2.0/hdf5-1.14.3-u6yn/bin/h5pfc -I.../gcc-13.2.0/hdf5-1.14.3-u6yn/include -L.../gcc-13.2.0/hdf5-1.14.3-u6yn/lib -lhdf5_fortran -lhdf5 -I/opt/cray/pe/mpich/8.1.28/ofi/gnu/10.3/include -I/opt/cray/pe/mpich/8.1.28/ofi/gnu/10.3/include -L/opt/cray/pe/mpich/8.1.28/ofi/gnu/10.3/lib -lmpifort_gnu_103 -lmpi_gnu_103
```
This seems fine. The full awk command would return:
```
$ .../gcc-13.2.0/hdf5-1.14.3-u6yn/bin/h5pfc -show | awk -F'-L' '{$1=""; for (i=2; i<=NF;i++) $i="-L"$i; print $0}'


 -L.../gcc-13.2.0/hdf5-1.14.3-u6yn/lib -lhdf5_fortran -lhdf5 -I/opt/cray/pe/mpich/8.1.28/ofi/gnu/10.3/include -I/opt/cray/pe/mpich/8.1.28/ofi/gnu/10.3/include  -L/opt/cray/pe/mpich/8.1.28/ofi/gnu/10.3/lib -lmpifort_gnu_103 -lmpi_gnu_103
```

Note the two blank line at the top. Because the result of the command gets injected into `make.inc` the \n chars also get into the file and we fail to use hdf5..

I use `xargs` to remove any blanc chars.

@bellenlau  @tgorni  @ye-luo

### CODE REVIEW

**Comment by ye-luo** [comment]
Could you make a fix to https://gitlab.com/QEF/q-e and then we can handle the patch in spack. In the meantime, choosing CMake should unblock you building QE using spack, right?

**Comment by etiennemlb** [comment]
QE's cmake build system is broken on my system:

```
     51    -- Looking for Fortran sgemm
     52    -- Looking for Fortran sgemm - not found
     53    -- Could NOT find BLAS (missing: BLAS_LIBRARIES)
     54    -- Could NOT find LAPACK (missing: LAPACK_LIBRARIES)
     55        Reason given by package: LAPACK could not be found because dependency BLAS could not be found.
     56
  >> 57    CMake Error at CMakeLists.txt:377 (message):
     58      Failed to find a complete set of external BLAS/LAPACK library by
     59      FindLAPACK.  Variables controlling FindLAPACK can be found at CMake online
     60      documentation.  Alternatively, '-DQE_LAPACK_INTERNAL=ON' may be used to
     61      enable reference LAPACK at a performance loss compared to optimized
     62      libraries.
```

Using autotools we explicitly tell QE where to find them, under QE's cmake it does not work. That said, I have plenty of other software building fine with cmake + lapack. My lapack provider is `cray-mpich`.

I may debug why, but later. Autotools works fine so that is enough for the moment.

**Comment by ye-luo** [comment]
> QE's cmake build system is broken on my system:
> 
> ```
>      51    -- Looking for Fortran sgemm
>      52    -- Looking for Fortran sgemm - not found
>      53    -- Could NOT find BLAS (missing: BLAS_LIBRARIES)
>      54    -- Could NOT find LAPACK (missing: LAPACK_LIBRARIES)
>      55        Reason given by package: LAPACK could not be found because dependency BLAS could not be found.
>      56
>   >> 57    CMake Error at CMakeLists.txt:377 (message):
>      58      Failed to find a complete set of external BLAS/LAPACK library by
>      59      FindLAPACK.  Variables controlling FindLAPACK can be found at CMake online
>      60      documentation.  Alternatively, '-DQE_LAPACK_INTERNAL=ON' may be used to
>      61      enable reference LAPACK at a performance loss compared to optimized
>      62      libraries.
> ```
> 
> Using autotools we explicitly tell QE where to find them, under QE's cmake it does not work. That said, I have plenty of other software building fine with cmake + lapack. My lapack provider is `cray-mpich`.
> 
> I may debug why, but later. Autotools works fine so that is enough for the moment.

I have limited experience using spack within Cray Programming Environment. It seems that `cray-mpich` bypasses Cray MPI wrappers (cc, CC, ftn). As a result, libsci won't be auto detected. You are responsible to specify a blas/lapack dependency. Maybe `^cray-libsci` or alternative BLAS/LAPACK libraries.

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 1]
```diff
@@ -333,6 +333,9 @@ class QuantumEspresso(CMakePackage, Package):
     patch_checksum = "72564c168231dd4a1279a74e76919af701d47cee9a851db6e205753004fe9bb5"
     patch(patch_url, sha256=patch_checksum, when="@6.7+qmcpack")
 
+    # 6.7
+    patch("configure.patch", when="@6.6:")
```
Comment (ye-luo) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py]: Please use `https://gitlab.com/QEF/q-e/-/commit/081409ea90cba0ddc07bea5ac29e3cd422c67d3d.diff` patch and also fix the comment `6.7` -> `6.6`. See lines above about how to use a commit as a patch .

**Review by ye-luo — CHANGES_REQUESTED** [review]
QE side has been settled.
Please finalize this PR.

File: var/spack/repos/builtin/packages/quantum-espresso/package.py (hunk 1 above)
Comment (etiennemlb) [reply var/spack/repos/builtin/packages/quantum-espresso/package.py]: Ok, merge when the CI passes.

**Review by ye-luo — APPROVED** [review]

**Review by tgamblin — APPROVED** [review]

File: var/spack/repos/builtin/packages/quantum-espresso/package.py (hunk 1 above)
Comment (ye-luo) [reply var/spack/repos/builtin/packages/quantum-espresso/package.py]: LGTM

**Pull request merged**